libs/cardano-ledger-core/src/Cardano/Ledger/UTxO.hs

```diff
@@ -0,0 +1,251 @@
+import Data.Sharing (FromSharedCBOR (Share, fromSharedCBOR), Interns)
+import Data.Typeable (Typeable)
+import GHC.Generics (Generic)
+import Lens.Micro ((^.))
+import NoThunks.Class (NoThunks (..))
+import Quiet (Quiet (Quiet))
+
+-- ===============================================
+
+-- | The unspent transaction outputs.
+newtype UTxO era = UTxO {unUTxO :: Map.Map (TxIn (EraCrypto era)) (TxOut era)}
+  deriving (Default, Generic, Semigroup)
+
+deriving instance NoThunks (TxOut era) => NoThunks (UTxO era)
+
+deriving instance (Era era, NFData (TxOut era)) => NFData (UTxO era)
+
+deriving newtype instance
+  (Eq (TxOut era), Crypto (EraCrypto era)) => Eq (UTxO era)
+
+deriving newtype instance Crypto (EraCrypto era) => Monoid (UTxO era)
+
+instance (Era era, ToCBOR (TxOut era)) => ToCBOR (UTxO era) where
+  toCBOR = encodeMap toCBOR toCBOR . unUTxO
+
+instance
+  ( Crypto (EraCrypto era),
+    FromSharedCBOR (TxOut era),
+    Share (TxOut era) ~ Interns (Credential 'Staking (EraCrypto era))
+  ) =>
+  FromSharedCBOR (UTxO era)
+  where
+  type
+    Share (UTxO era) =
+      Interns (Credential 'Staking (EraCrypto era))
+  fromSharedCBOR credsInterns =
+    UTxO <$!> decodeMapNoDuplicates fromCBOR (fromSharedCBOR credsInterns)
+
+instance
+  ( FromCBOR (TxOut era),
+    Era era
+  ) =>
+  FromCBOR (UTxO era)
+  where
+  fromCBOR = UTxO <$!> decodeMapNoDuplicates fromCBOR fromCBOR
+
+deriving via
+  Quiet (UTxO era)
+  instance
+    (Show (TxOut era), Crypto (EraCrypto era)) => Show (UTxO era)
+
+-- | Compute the UTxO inputs of a transaction.
+-- txins has the same problems as txouts, see notes below.
+txins ::
+  EraTxBody era =>
+  TxBody era ->
+  Set (TxIn (EraCrypto era))
+txins = (^. inputsTxBodyL)
+
+-- | Compute the transaction outputs of a transaction.
+txouts ::
+  forall era.
+  EraTxBody era =>
+  TxBody era ->
+  UTxO era
+txouts txBody =
+  UTxO $
+    Map.fromList
+      [ (TxIn transId idx, out)
+        | (out, idx) <- zip (toList $ txBody ^. outputsTxBodyL) [minBound ..]
+      ]
+  where
+    transId = txid txBody
+
+-- | Lookup a txin for a given UTxO collection
+txinLookup ::
+  TxIn (EraCrypto era) ->
+  UTxO era ->
+  Maybe (TxOut era)
+txinLookup txin (UTxO utxo') = Map.lookup txin utxo'
+
+-- | Verify a transaction body witness
+verifyWitVKey ::
+  ( Typeable kr,
+    Crypto c,
+    DSignable c (Hash c EraIndependentTxBody)
+  ) =>
+  Hash c EraIndependentTxBody ->
+  WitVKey kr c ->
+  Bool
+verifyWitVKey txbodyHash (WitVKey vkey sig) = verifySignedDSIGN vkey txbodyHash (coerce sig)
+
+-- | Create a witness for transaction
+makeWitnessVKey ::
+  forall c kr.
+  ( Crypto c,
+    DSignable c (CH.Hash (HASH c) EraIndependentTxBody)
+  ) =>
+  SafeHash c EraIndependentTxBody ->
+  KeyPair kr c ->
+  WitVKey 'Witness c
+makeWitnessVKey safe keys =
+  WitVKey (asWitness $ vKey keys) (coerce $ signedDSIGN @c (sKey keys) (extractHash safe))
+
+-- | Create witnesses for transaction
+makeWitnessesVKey ::
```

Comment:
this is just a side comment while I'm thinking about it, definitely ignore for now in this PR. but I think lots of the functions in this module are not really that related to the `UTxO` and should be in their own module.

Reply:
For the next cleanup PR